src/util/std_code.h

```diff
@@ -16,8 +16,17 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "expr.h"
 #include "expr_cast.h"
 
-/*! \brief A statement in a programming language
-*/
+/// Data structure for representing an arbitrary statement in a program. Every
+/// specific type of statement (e.g. block of statements, assignment,
+/// if-then-else statement...) is represented by a subtype of `codet`.
+/// `codet`s are represented to be subtypes of \ref exprt since statements can
+/// occur in an expression context in C: for example, the assignment `x = y;`
+/// is an expression with return value `y`. For other types of statements in an
+/// expression context, see e.g.
+/// https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html
```

Comment:
True, but I'm wondering whether today that's a good design decision. We might end up with better compiler diagnostics if we made `codet` inherit from `irept` directly. If anyone else thinks the same, would they mind creating an issue suggesting we investigate this?

Reply:
Yes, this might be an option.

Reply:
I actually got confused about this part and raised the question "Why are we making codet inherit from exprt?" with several people at Diffblue. This part of the documentation is a summary of the answers I got. I still find it a bit confusing that codet inherits from exprt, but side_effect_exprt does not inherit from codet and instead duplicates parts of codet. Investigating if the hierarchy could be improved sounds good to me - if it's not too much effort at this point.

Reply:
#2810